### PR TITLE
replace fs implementation with v3 client

### PR DESCRIPTION
**Pull request by jhsinger-klotho**
• Does any part of it require special attention?
• Does it relate to or fix any issue? closes https://github.com/klothoplatform/klotho-history/issues/528

Changing the fs implementation in node so we dont use the large aws sdk. also added delete object as unlink

Tested by using the cloudfs sample app and running integration tests

### Standard checks

- **Unit tests**: Any special considerations? no
- **Docs**: Do we need to update any docs, internal or public? upon approval will add delete to docs in node api
- **Backwards compatibility**: Will this break existing apps? If so, what would be the extra work required to keep them working? yes

### CODE REVIEW

File: pkg/lang/javascript/aws_runtime/_fs.ts [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+const streamToString = (stream: Readable): Promise<string> =>
+    new Promise((resolve, reject) => {
+        const chunks: Uint8Array[] = []
+        stream.on('data', (chunk) => chunks.push(chunk))
+        stream.on('error', reject)
+        stream.on('end', () => resolve(Buffer.concat(chunks).toString('utf8')))
+    })
+
+async function getCallParameters(paramKey, dispatcherMode) {
+    let isEmitter = dispatcherMode === 'emitter' ? true : false
```
Comment (yuval-klotho) [inline pkg/lang/javascript/aws_runtime/_fs.ts:31]: nit:
```suggestion
    let isEmitter = dispatcherMode === 'emitter'
```

File: pkg/lang/javascript/aws_runtime/_fs.ts [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+
+        let parameters = ''
+        if (result.Body) {
+            parameters = await streamToString(result.Body as Readable)
+        }
+
+        if (isEmitter && Array.isArray(parameters)) {
+            // Emitters only have 1 parameter - the runtime saves an array, so we
+            // normalize the parameter
+            parameters = _.get(parameters, '[0]')
```
Comment (yuval-klotho) [inline pkg/lang/javascript/aws_runtime/_fs.ts:47]: How is this different from just `parameters[0]`? I tested, and it looks like `parameters[0]` evaluates to undefined if the array is empty, so we don't need index-bounds protection.

File: pkg/lang/javascript/aws_runtime/_fs.ts [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+            if (Array.isArray(parameters)) {
+                let paramPairs = _.toPairs(parameters)
+                paramPairs = paramPairs.map((x) => {
+                    if (_.get(x, '[1].type') == 'Buffer') {
```
Comment (yuval-klotho) [inline pkg/lang/javascript/aws_runtime/_fs.ts:51]: similarly, if this is for index bounds, I _think_ typescript can do it as just `if (x[1]?.type == 'Buffer')`

File: pkg/lang/javascript/aws_runtime/_fs.ts [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+                    }
+                })
+
+                parameters = _.fromPairs(paramPairs)
+            }
+        }
+
+        return parameters || {}
+    } catch (e) {
+        console.error(e)
```
Comment (yuval-klotho) [inline pkg/lang/javascript/aws_runtime/_fs.ts:64]: Why don't we want to rethrow? (Ditto for similar lines below)

File: pkg/lang/javascript/aws_runtime/_fs.ts [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+
+async function s3_writeFile(...args) {
+    const bucketParams = {
+        Bucket: payloadBucketPhysicalName,
+        Key: `${userBucketPath}/${args[0]}`,
+        Body: args[1],
+    }
+    try {
+        await s3Client.send(new PutObjectCommand(bucketParams))
+        console.log('Successfully uploaded object: ' + bucketParams.Bucket + '/' + bucketParams.Key)
```
Comment (yuval-klotho) [inline pkg/lang/javascript/aws_runtime/_fs.ts:93]: I wonder if this should be debug? I'd be worried about spamming the service logs with what is essentially an implementation detail (they should be able to assume that if the s3 didn't log a failure, that it succeeded).

Similar for other log lines, other than the ones within catches.

**Review by yuval-klotho — APPROVED** [review]
I had a few questions as I looked at the added lines, but then when I looked at the removed lines, it looks like this is pretty equivalent to what was there before. So, treat all of my comments as non-blocking — my approval is intended. :)

File: pkg/lang/javascript/aws_runtime/_fs.ts (hunk 2 above)
Comment (jhsinger-klotho) [reply pkg/lang/javascript/aws_runtime/_fs.ts:47]: a lot of this was copy paste from the old implementation, going to try to remove some of the more odd things like this

File: pkg/lang/javascript/aws_runtime/_fs.ts (hunk 4 above)
Comment (jhsinger-klotho) [reply pkg/lang/javascript/aws_runtime/_fs.ts:64]: good call, we should ill add that in, was my miss

File: pkg/lang/javascript/aws_runtime/_fs.ts (hunk 5 above)
Comment (jhsinger-klotho) [reply pkg/lang/javascript/aws_runtime/_fs.ts:93]: hm, yeah i can add them all as debug

File: pkg/lang/javascript/aws_runtime/_fs.ts [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+        parameters = JSON.parse(parameters)
+
+        if (isEmitter && Array.isArray(parameters)) {
+            // Emitters only have 1 parameter - the runtime saves an array, so we
+            // normalize the parameter
+            parameters = parameters[0]
+            if (Array.isArray(parameters)) {
+                let paramPairs = _.toPairs(parameters)
+                paramPairs = paramPairs.map((x) => {
+                    if (x[1].type == 'Buffer') {
```
Comment (gordon-klotho) [inline pkg/lang/javascript/aws_runtime/_fs.ts:51]:
```suggestion
                    if (x[1] instanceof Buffer) {
```

File: pkg/lang/javascript/aws_runtime/_fs.ts [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+                        return [x[0], Buffer.from(x[1].data)]
```
Comment (gordon-klotho) [inline pkg/lang/javascript/aws_runtime/_fs.ts:52]: Is this supposed to be a clone then? It's already a buffer, so this just reads it into another buffer, right?

File: pkg/lang/javascript/aws_runtime/_fs.ts [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+            parameters = await streamToString(result.Body as Readable)
+        }
+        parameters = JSON.parse(parameters)
+
+        if (isEmitter && Array.isArray(parameters)) {
+            // Emitters only have 1 parameter - the runtime saves an array, so we
+            // normalize the parameter
+            parameters = parameters[0]
+            if (Array.isArray(parameters)) {
+                let paramPairs = _.toPairs(parameters)
```
Comment (gordon-klotho) [inline pkg/lang/javascript/aws_runtime/_fs.ts:49]: `Object.entries` I believe is the equivalent to `toPairs`

**Review by gordon-klotho — COMMENTED** [review]
1 last usage of lodash we can remove

File: pkg/lang/javascript/aws_runtime/_fs.ts (hunk 7 above)
Comment (jhsinger-klotho) [reply pkg/lang/javascript/aws_runtime/_fs.ts:56]: looks like it, i can just change it to return x[1].data

File: pkg/lang/javascript/aws_runtime/_fs.ts (hunk 8 above)
Comment (jhsinger-klotho) [reply pkg/lang/javascript/aws_runtime/_fs.ts:53]: 👍

**Pull request merged**